aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java

```diff
@@ -128,7 +128,7 @@ public Orchestrator(
                 .queryPredicateProvider(queryPredicateProvider)
                 .onFailure(this::onApiSyncFailure)
                 .build();
-        this.storageObserver = new StorageObserver(localStorageAdapter, mutationOutbox);
+        this.storageObserver = new StorageObserver(localStorageAdapter, mutationOutbox, queryPredicateProvider);
```

Comment:
IMO, the `StorageObserver` is just a dumb little utility to watch the DataStore and funnel stuff into the `MutationOutbox`.

Instead of giving the StorageObserver awareness of the filtering rules, how about giving this responsibility to the  **`MutationProcessor`**?

When the MutationProcessor is evaluating a mutation, it can say "whoa! Hold the phone. I'm not going to send _this_, are you kidding me? It's sync predicate is all wrong."

----

Related question: Are sync predicates for outgoing mutations, today? Or just inbound?

Seems actually like this is not a unique problem to the `.none()` filter?

Reply:
> Related question: Are sync predicates for outgoing mutations, today? Or just inbound?

No, they are not applied at all for mutations today.   Today, sync predicates are applied to the sync query (on the server side), and subscriptions (today, on the client side, but in the future, server side, when AppSync support it)

> Seems actually like this is not a unique problem to the .none() filter?

This is a really good point that I've been thinking about.  This PR only filters mutations if the syncExpression is set to `.none()`.  We really shouldn't treat `.none()` as a special case though.  We have a design tenet that going something like, "Before adding support for a special case, think twice, because you probably don't _really_ need it."

Then the question becomes - should we apply all sync expressions to mutations, or not?     I actually think we should _not_.  The primary use case for sync expressions is to speed up the sync process and reduce data usage by reducing the total data that is actually synced.   Customers should not expect all of the data in DataStore to evaluate to true against the sync expressions.  For example, a customer starts DataStore with syncExpression A, then stops and starts again with syncExpression B, the local database will contain the union of A & B.  It is not intended to be used as an implicit filter on `DataStore.query`, even though it may sometimes work that way.  `DataStore.query` takes a `QueryPredicate` specifically for that purpose.

The one reason _to_ apply mutations to sync expressions is to support the use case where a customer wants some of their models to sync with their API, and others to just use to save data locally.  I don't think we should invest in supporting this use case though, because it feels outside the scope of Amplify. Customers wanting to save data locally should just use their own SQLite DB, or any other Android library for local persistence.

TL;DR: I'm going to remove the mutation filtering from this PR.

Reply:
Sounds good! Another (rather hairy) aspect to this:

If we don't do a server-to-device sync for some models, we won't know their correct version. Now, let's say we *do* allow device-to-server sync ("mutations") for those same models. The mutations will go out with incorrect versions. In the best case (optimistic concurrency) the mutations will fail with an error and the conflict handler will be invoked. But in the 80% case (Auto Merge), the mutation will fail silently and we will lose customer data without any error being raised.

Reply:
Hmm, that is a good point.  I believe that scenario is a problem today, irrespective of this PR. 

1. Client 1 starts DataStore with sync predicate X
2. Client 1 creates item A, version 1, which does **not** match sync predicate X.
3. Client 2 receives item A, version 1, via a subscription.
4. Client 2 updates item A, incrementing `_version` to 2.
5. Client 1 now has stale version of item A,  _version 1
5. Client 1 updates item A, version 1.  For Auto Merge, the mutation will fail silently, and we will lose customer data.

To solve this, maybe we _should_ actually have a filter in the `MutationProcessor`?  And if the mutation doesn't match the sync predicate, then, we could fire an `outboxMutationFailed` event, containing the item, to prevent customer data loss.  Currently, `outboxMutationFailed` is only fired for API errors, but perhaps we could also use it for this use case.

Reply:
That sounds potentially like a mitigation, yes. Maybe we should also talk about the silent Auto Merge failure w/ the DataStore Working Group / Jonny P? I'm worried that it may be causing errors that we're not even aware of -- although, we've identified a few failure modes, to date. Perhaps we can optionally turn on "strict version matching" via a request header or something?